### PR TITLE
feat(core): add application-specific custom banners

### PR DESCRIPTION
**Pull request by maggieneterval**
Let me know if you all have any thoughts on making the custom banners more or less configurable -- the specific use case this was requested for was putting up production freeze or maintenance alerts on specific applications.

b5455002-3461-4792-b8f9-c8e06eae1b1f

### CODE REVIEW

**Comment by bpowell** [comment]
Awesome feature! We had something like this on the application I used to work on, it was extremely helpful. 

Is there anyway to add a link to documentation or something so people know why the freeze/maintenance is happening?

**Comment by bpowell** [comment]
After talking to @ethanfrogers about this, a cool idea we came up with would be to have an emergency alert system. For example, PageDuty is alerting about a system being down, have that reflect in the Spinnaker UI for that application. I'd say this is more of a v2 or v3 feature for this, but figured I should bring it up

**Comment by anotherchrisberry** [comment]
This looks really cool.

Could you also look at the application's attributes from Front50 to see if a banner is configured there? Not sure what the UI would look like, maybe another section on the config tab. But this would give application owners some control over the banners, and avoid having to redeploy Deck to make changes to banners.

**Comment by gcomstock** [comment]
Looks nice! Just make sure the extra height/offset is compatible with the blesk messaging, and doesn't push the scrollable sections down and cause cropping.

**Comment by maggieneterval** [comment]
Thanks all for the feedback! Adding links and automatic alert functionality sound like a good v2 for this -- hopefully we get some user feedback to inform what might be most helpful.

@anotherchrisberry I like the idea of moving the banner configuration out of settings and into an application's Front50 attributes -- I'll put some thought into the UI for this on the config tab and post another screenshot once I've updated this PR.

**Comment by maggieneterval** [comment]
@gcomstock naive question for you -- what is the blesk messaging?

**Comment by gcomstock** [comment]
> @gcomstock naive question for you -- what is the blesk messaging?

Ah Sorry Maggie, I've been informed its a Netflix-specific banner that also goes across the top and pushes the page down.

**Comment by maggieneterval** [comment]
Just pushed an update to support configuring application banners from the Config tab -- let me know what you think of the UI. After some discussion here we decided it would be best to limit users to several Spinnaker color palette options for the banner text and background color.

@anotherchrisberry or @sbwsg this is an angular newb question -- is there a preferred method in Deck for communicating among components without a close ancestor, so that we can trigger an update in the header after a new banner config is enabled? Currently the user would need to refresh the page for an updated banner to take effect. Thanks!

**Comment by jtk54** [comment]
@maggieneterval is this something we expect to be able to control via `spin app save`?

**Comment by jtk54** [comment]
Actually judging from [what's there](https://github.com/spinnaker/spin/blob/master/cmd/application/save.go) so far users will be able to submit app banner config as long as it's discoverable somehow. I think at the moment folks can inspect the traffic to see the payload sent but it might be worth documenting or adding a 'See application json' similarly to pipelines.

File: app/scripts/modules/core/src/application/config/customBanner/CustomBannerConfig.tsx [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+                  </td>
+                </tr>
+              ))}
+            </tbody>
+          </table>
+        </div>
+        <div className="col-md-10 col-md-offset-1">
+          <button className="btn btn-block add-new" onClick={this.addBanner}>
+            <span className="glyphicon glyphicon-plus-sign" />
+            Add banner
```
Comment (anotherchrisberry) [inline app/scripts/modules/core/src/application/config/customBanner/CustomBannerConfig.tsx]:
```suggestion
            {' '}Add banner
```
should put a space between the icon and the text

**Comment by anotherchrisberry** [comment]
I think this looks great! I'm not sure if you could use the [formik forms stuff](https://github.com/spinnaker/deck/tree/master/app/scripts/modules/core/src/presentation/forms) that @christopherthielen and @alanmquach have been working on, but that might also be a nice TODO later.

File: app/scripts/modules/core/src/application/config/customBanner/CustomBannerConfig.tsx (hunk 1 above)
Comment (maggieneterval) [reply app/scripts/modules/core/src/application/config/customBanner/CustomBannerConfig.tsx]: Thanks Chris, just updated!

**Comment by emagana** [comment]
@maggieneterval  Is there a way to set-up a Banner but not at application level only but for the main top section of the UI where we can differentiate a Dev Spinnaker instance from a PROD one?

**Comment by maggieneterval** [comment]
@emagana Not yet, the current implementation is only per-application!

**Comment by emagana** [comment]
Thanks for the quick response. I think I should use slack, next time.  :-)

**Pull request merged**